Add a tutorial doc for composition revisions

ezgidemirel opened the pull request:
Signed-off-by: ezgidemirel <ezgidemirel91@gmail.com>

<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes
Adds a tutorial doc for Composition Revisions. This is quite different than what we have in the guides folder. The idea is to provide some commands to apply for trying the feature. 

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested
Applied the commands and compared them with the expected outputs. 
<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

plumbis commented:
I'm going to add a Knowledge Base section to the docs before merging this in

plumbis commented:
I tried a few times to submit my edits directly and failed.  I put them in a commit on my personal fork.
https://github.com/plumbis/crossplane/commit/86ef37afc7aa64ba4671adee5c48fa8a30f4b115

The edits are mainly removing [passive voice](https://www.grammarly.com/blog/passive-voice/) and adding improved styling. I also removed the `kubectl apply -f` blocks since there has been disagreement about what we should standardize on in the docs. 

I moved this into the `/guides` directory and we'll find a better home for it after the upcoming docs reorg. 
As a result we also need to [apply a `weight`](https://github.com/plumbis/crossplane/commit/86ef37afc7aa64ba4671adee5c48fa8a30f4b115#diff-be7bd746bfeae28b06d16ba11d7e31c64157c24299b7fb615696ff3d7c929c02R3) setting to the existing composition page.

turkenh commented on docs/guides/composition-revisions-example.md:
```diff
@@ -0,0 +1,302 @@
+This tutorial creates multiple XRs and modifies Composition to apply different Composition Revisions to the XRs.
+
+## Install Crossplane
+1. Install Crossplane v1.11.0 or later.
```
```suggestion
1. Install Crossplane v1.11.0 or later and wait until the Crossplane pods are running.
```

turkenh commented on docs/guides/composition-revisions-example.md:
````diff
@@ -0,0 +1,302 @@
+crossplane-rbac-manager-78bd597746-sdv6w   1/1     Running   0          9s
+```
+
+## Apply a Composition
````
What do you think about making this step 2 of a Preparation section? 
I believe it is easier when steps are grouped as much as possible and presented as numbered steps to follow.

turkenh commented on docs/guides/composition-revisions-example.md:
````diff
@@ -0,0 +1,302 @@
+myvpcs.aws.example.upbound.io-ad265bc   1          dev
+```
+
+The label `dev` is automatically created from the Composition.
````
Should the column title be CHANNEL instead?

And with something like:
```suggestion
All labels propagated from `Composition` to the `CompositionRevision` including the `channel` label, which we printed as a column using the get query we built above.
```

ezgidemirel replied:
updated as CHANNEL

turkenh commented on docs/guides/composition-revisions-example.md:
```diff
@@ -0,0 +1,302 @@
+The label `dev` is automatically created from the Composition.
+
+## Create Composite Resources
+
```
A short intro about what we will be doing in this section and what we will see at the end would really be helpful.

ezgidemirel replied:
done

turkenh commented on docs/guides/composition-revisions-example.md:
````diff
@@ -0,0 +1,302 @@
+Verify the Composite Resource with the label `channel: staging` doesn't have a `REVISION`.  
+All other XRs have a `REVISION` matching the created Composition Revision.
+```shell
+kubectl get composite -o=custom-columns=NAME:.metadata.name,SYNCED:.status.conditions[0].status,REVISION:.spec.compositionRevisionRef.name,POLICY:.spec.compositionUpdatePolicy,MATCHLABEL:.spec.compositionRevisionSelector.matchLabels
````
Using zsh and this command failed as is: 

```
zsh: no matches found: -o=custom-columns=NAME:.metadata.name,SYNCED:.status.conditions[0].status,REVISION:.spec.compositionRevisionRef.name,POLICY:.spec.compositionUpdatePolicy,MATCHLABEL:.spec.compositionRevisionSelector.matchLabels
```
It worked fine, after I quote as follows:

```sh
kubectl get composite -o="custom-columns=NAME:.metadata.name,SYNCED:.status.conditions[0].status,REVISION:.spec.compositionRevisionRef.name,POLICY:.spec.compositionUpdatePolicy,MATCHLABEL:.spec.compositionRevisionSelector.matchLabels"
```

turkenh commented on docs/guides/composition-revisions-example.md:
````diff
@@ -0,0 +1,302 @@
+```
+Expected Output:
+```shell
+NAME          SYNCED   REVISION                                POLICY      MATCHLABEL
````
Do you think if printing CHANNEL directly would make more sense?

ezgidemirel replied:
I couldn't index the channel label in the command 🙈

turkenh commented on docs/guides/composition-revisions-example.md:
```diff
@@ -0,0 +1,302 @@
+
+{{< hint "note" >}}
+The `vpc-staging` XR label doesn't match any existing Composition Revisions.
+{{< /hint >}}
```
It looks like this is not rendering as expected.

<img width="933" alt="Screen Shot 2022-12-15 at 14 49 27" src="https://user-images.githubusercontent.com/9900707/207852015-732d9b5d-d4b0-4acb-9311-eed5b8fa5f87.png">

ezgidemirel replied:
this will be fixed on the docs site, please see the `Hints and alert boxes` chapter [here](https://docs.crossplane.io/master/contributing/docs/#code-of-conduct).

turkenh commented on docs/guides/composition-revisions-example.md:
````diff
@@ -0,0 +1,302 @@
+``` 
+
+## Verify the Composite Resources
+Verify the Composite Resource with the label `channel: staging` doesn't have a `REVISION`.  
````
```suggestion
Verify the Composite Resource with the label `channel: staging` doesn't have a `REVISION` since ...
```

ezgidemirel replied:
I think explaining this in the hint below is better, just to give the reader a chance to think about it

turkenh commented on docs/guides/composition-revisions-example.md:
````diff
@@ -0,0 +1,302 @@
+XRs `vpc-man` and `vpc-dev` are still assigned to the original revision:1:
+
+```shell
+kubectl get composite -o=custom-columns=NAME:.metadata.name,SYNCED:.status.conditions[0].status,REVISION:.spec.compositionRevisionRef.name,POLICY:.spec.compositionUpdatePolicy,MATCHLABEL:.spec.compositionRevisionSelector.matchLabels
````
same issue with zsh here.

turkenh commented on docs/guides/composition-revisions-example.md:
````diff
@@ -0,0 +1,302 @@
+vpc-staging   True     myvpcs.aws.example.upbound.io-727b3c8   Automatic   map[channel:staging]
+``` 
+
+{{< hint "note" >}}
````
Same issue with rendering here.

ezgidemirel replied:
this will be fixed on the docs site, please see the `Hints and alert boxes` chapter [here](https://docs.crossplane.io/master/contributing/docs/#code-of-conduct).

plumbis replied:
You should be able to run `make docs.run` to render the page locally. Because of how versions work against a branch name, the page will be at http://localhost:1313/comp-rev-tutorial/guides/composition-revisions-example/

turkenh commented on docs/guides/composition-revisions-example.md:
```diff
@@ -0,0 +1,302 @@
+`vpc-staging` now matches the label applied to Revision revision:2.
+{{< /hint >}}
+
+### Enable DNS support in the Composition
```
Having this as a title feels confusing TBH. 
`Updating Composition` would be more relevant.

turkenh commented on docs/guides/composition-revisions-example.md:
````diff
@@ -0,0 +1,302 @@
+myvpcs.aws.example.upbound.io-ad265bc   1          dev
+myvpcs.aws.example.upbound.io-f81c553   3          dev
+``` 
+
````
What happened above? A short explanation would be helpful.
The explanation below should be moved after user see output of the next command.

ezgidemirel replied:
done

turkenh reviewed (commented):
Thanks @ezgidemirel. 
I could successfully go through the tutorial with no significant issues. Left a couple of comments, mostly around giving some more context on the steps to make it more digestible.

plumbis commented on docs/guides/composition-revisions-example.md:
```diff
@@ -0,0 +1,302 @@
+---
+This tutorial discusses how CompositionRevisions work and how they manage Composite Resource
+(XR) updates. This start with a `Composition` and `CompositeResourceDefinition` that defines a `MyVPC`
+resource.
```
@ezgidemirel if you want to add this I would put it after these lines. Start with a straight forward "what this covers" (that's what is here) followed by the "how it's done" (What @turkenh is asking for). 

Also try and avoid "we" (or better yet just make a statement without you/we). https://developers.google.com/style/person

plumbis replied:
Also I just saw that this line says `This start with` and should be _starts_ (+s).

ezgidemirel commented:
@turkenh ready for the second pass.

turkenh commented:
> @turkenh ready for the second pass.

I would like to highlight my following comment which seems to have resolved yet:

> I believe it is easier when steps are grouped as much as possible and presented as numbered steps to follow.

As we discussed offline, I believe it is much easier to follow a tutorial when there are numbered steps grouped as logical units. Something like:

```
1. Preparation
1.1. Install Crossplane
1.2. Deploy Composition and CompositeResourceDefinition

2. Creating Composites
2.1. With default update policy
2.2. With manual update policy
2.3. Using a revision selector
2.4. Verify the Composites

3. Updating Composition
3.1. Update the Composition label
3.2. Update the Composition Spec
3.3. Verify Revisions of Composites
```

@ezgidemirel @plumbis what do you think?

plumbis commented:
I would recommend against the numbers and allow the headers to be the flow. The right-hand table of contents automatically provides the nesting based on headers. 
![Screen Shot 2022-12-16 at 9 21 24 AM](https://user-images.githubusercontent.com/10537576/208118830-285d6b2f-f717-414f-a823-182caaf2bd86.png)

turkenh commented:
Ok, I still believe it is easier to follow with numbers as [here](https://docs.divio.com/en/latest/introduction/aldryn/aldryn-django-02-create-project/), however, I just checked https://kubernetes.io/docs/tutorials/ and noticed that they didn't use numbering either.

Maybe tweaking headers as something like I proposed (but without numbers) might help understand the flow.

ezgidemirel commented:
I think we had too many headers for such a small tutorial. I removed the `verify` ones and changed the commands in each section to list items. @turkenh @plumbis what do you guys think?

turkenh reviewed (approved):
Looking great, thanks @ezgidemirel!
